chore(bigquery): Add extra logging for BigQuery exceptions so we can have better insight on exceptions

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
Some BigQuery exceptions are not being logged with enough information, so we need to add some extra method that parses the original exception and logs the correct message. We also add some unit tests to the new method.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
After:
![test](https://user-images.githubusercontent.com/38889534/199721944-5717e9b9-698c-406b-90c6-0b4a4bd4c2a3.png)


### TESTING INSTRUCTIONS
1. Connect Locally to a BigQuery DB
2. Run an invalid SQL statement so you get a BigQuery exception
3. Check you are logging the correct message out of the original exception

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
BigQuery:

- Add new method to parse unknown exception in BigQuery engines. This method is going to shed some light so we can show correct messages later using our regex

hughhhh commented on superset/db_engine_specs/base.py:
```diff
@@ -443,6 +452,9 @@ def get_dbapi_mapped_exception(cls, exception: Exception) -> Exception:
         new_exception = cls.get_dbapi_exception_mapping().get(type(exception))
         if not new_exception:
+            # We are interested in just BigQuery exceptions
+            if cls.engine == "bigquery":
```
can we add `bigquery` to this engine + add it here as `Engines.BigQuery`

https://github.com/apache/superset/blob/66c0801e494f16ba44d492f3648c1e1a57c8e35a/superset-frontend/src/views/CRUD/data/database/types.ts#L150

Antonio-RiveroMartnez replied:
Hmm, that's in the `superset-frontend` module, right? Do we have some similar enum in Python?

hughhhh commented on superset/db_engine_specs/bigquery.py:
```diff
@@ -578,3 +578,7 @@ def _get_fields(cls, cols: List[Dict[str, Any]]) -> List[Any]:
+
+    @classmethod
+    def parse_error_exception(cls, exception_message: str) -> str:
+        return exception_message.rsplit("\n")[0]
```
Can we have better try/exception catching here just incase we can't find the new line? Then a write test to back this whenever `parse_error_exception` raises.

Antonio-RiveroMartnez committed:
BigQuery:

    - Use new EngineType enum instead of raw string
    - Add better exception handling for new parse_error_exception method in case anything fails during the parsing
    - Add test to back the exception handling

hughhhh commented on superset/db_engine_specs/base.py:
```diff
@@ -443,6 +452,9 @@ def get_dbapi_mapped_exception(cls, exception: Exception) -> Exception:
         new_exception = cls.get_dbapi_exception_mapping().get(type(exception))
         if not new_exception:
+            # We are interested in just BigQuery exceptions
+            if cls.engine == EngineType.BIGQUERY:
```
a better check is if the class has the function `parse_error_exception`

Antonio-RiveroMartnez replied:
They all have it because the base class have it, I had it that way initially but our hooks and linting rules don't like it because `cls` would be the base class and it checks with it instead of the BigQuery one for example. So, Was that or ignoring the rule if possible.

hughhhh commented on superset/utils/core.py:
```diff
@@ -190,6 +190,10 @@ class DatasourceType(str, Enum):
     VIEW = "view"
 
 
+class EngineType(str, Enum):
```
👍🏾

eschutho reviewed (changes requested):
look at inheritance.. come back to this

Antonio-RiveroMartnez committed:
BigQuery:

- parse_error_exception returns an exception with the parsed message, because of that we can skip the compare with Bigquery and instead just make use of the existing logger upstream (sql_lab)

hughhhh commented on superset/db_engine_specs/base.py:
```diff
@@ -430,6 +430,15 @@ def get_dbapi_exception_mapping(cls) -> Dict[Type[Exception], Type[Exception]]:
+
+        :return: An Exception with a parsed string off the original exception
+        """
+        return exception
```
make this NotImplemented()

Antonio-RiveroMartnez commented on tests/unit_tests/db_engine_specs/test_bigquery.py:
```diff
@@ -244,3 +244,74 @@ def test_mask_encrypted_extra_when_empty() -> None:
+    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |
+
+
+    bigquery error: 400 Table \"case_detail_all_suites\" must be qualified with a dataset (e.g. dataset.table).
```
Updated 👍

Antonio-RiveroMartnez committed:
BigQuery:

- Update our method and tests so bigquery error: portion doesn't appear in the message

eschutho commented on superset/db_engine_specs/bigquery.py:
```diff
@@ -578,3 +578,17 @@ def _get_fields(cls, cols: List[Dict[str, Any]]) -> List[Any]:
+        try:
+            return Exception(
+                str(exception)  # pylint: disable=use-maxsplit-arg
+                .rsplit("\n")[0]
```
you can also use `.splitlines()[0]` here

eschutho commented on superset/db_engine_specs/bigquery.py:
```diff
@@ -578,3 +578,17 @@ def _get_fields(cls, cols: List[Dict[str, Any]]) -> List[Any]:
+                .rsplit(":")[1]
+                .strip()
+            )
+        except Exception:  # pylint: disable=broad-except
```
I tried to get this to raise an exception and wasn't able to. I believe as long as the return value is a string, you should be able to perform all of the operations and don't need to put it in a try catch.

Antonio-RiveroMartnez replied:
Hmm if for some unknown reason the Exception has no `:` for example, accessing the [1] would cause a: `IndexError: list index out of range`.  Just adding the try/catch there in case anything like that happens with the message we get.

Antonio-RiveroMartnez committed:
BigQuery:

- Use splitlines instead and avoid ignoring pylint rules

eschutho reviewed (approved):
This looks great @Antonio-RiveroMartnez!
